analysis/token/synonym/synonym.go

```diff
@@ -0,0 +1,386 @@
+//  Copyright (c) 2023 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package synonym
+
+import (
+	"log"
+	"sync"
+
+	"github.com/blevesearch/bleve/v2/analysis"
+	"github.com/blevesearch/bleve/v2/registry"
+	"github.com/blevesearch/bleve/v2/search"
+	"github.com/blevesearch/vellum"
+)
+
+const Name = "synonym"
+
+const SeparatingCharacter = ' '
+
+type SynonymFilter struct {
+	fst            []byte
+	hashToSynonyms map[uint64][]uint64
+	hashToPhrase   map[uint64][]byte
+	fuzziness      int
+	prefix         int
+}
+
+// fstPath is used for storing the data associated
+// with a particular path in the FST. This would be used for enabling
+// fuzzy search in the synonym filter.
+type fstPath struct {
+	// output is the sum of the outputs of all the states seen in the path
+	output uint64
+	// state is the state at the end of the path
+	state int
+	// word is the sequence of characters that were matched in the path
+	word []byte
+}
+
+func resetPath(path *fstPath) *fstPath {
+	path.output = 0
+	path.state = 0
+	path.word = nil
```

Comment:
rather than assigning to nil, i'd recommend resizing this slice to a length 0 slice. path.word = path.word[:0]. that way the backing array won't be GC'd and fresh allocation won't happen too much.